api/types.go

```diff
@@ -1231,7 +1232,43 @@ type DeleteOrganizationMembershipPayload struct {
 }
 
 type MachineEvent struct {
-	Kind      string      `json:"kind"`
-	Timestamp time.Time   `json:"timestamp"`
-	Body      interface{} `json:"body"`
+	ID        string
+	Kind      string
+	Timestamp time.Time
+	ExitCode  *int
+	OOMKilled bool
+	// Body      map[string]interface{}
+}
+
+func (e *MachineEvent) UnmarshalJSON(data []byte) error {
+	raw := map[string]interface{}{}
+
+	if err := json.Unmarshal(data, &raw); err != nil {
```

Comment:
I think unmarshalling the data directly into the struct would save a lot of trouble when it comes to type assertions, something like:

```
var raw MachineEvent
if err := json.Unmarshal(data, &raw); err != nil {
  return err
}
```

Which will then allow you to just directly check for data like raw.ID etc

Reply:
MachineEvent takes more than one form such as `MachineEventStart`, `MachineEventExit`... and they don't share all the fields and so I must be more generic and manually check if the fields exists before trying to set them.

Reply:
On second thought I think a case switch could work, something like this: https://play.golang.com/p/49xJe_HysXb